### DISCUSSION TITLE
Story decorators using providers are not recognised when imported to use with React Testing Library

### DISCUSSION THREAD

**Discussion by daniel-norris** [storybookjs/storybook#25145] [Help] [answered]
### Summary

I'm refactoring a load of our Storybook files and tests and wanted to make use of the `decorators` at the story level but unsure if I am setting them up incorrectly, if it isn't designed to be used this way or if this is a bug. 

Here's a simplified example of what I'm doing: 

```typescript
// Button.stories.tsx

export const withProvider =
  (state: State): DecoratorFn =>
  (Story: ComponentStory<any>) =>
    (
      <Provider store={getStore(state)}>
        <Story />
      </Provider>
    );

export default {
  title: "Button",
  component: Button,
  decorators: [
    withProvider(getInitialState() as State),
  ],
} as ComponentMeta<typeof Button>;

const Template: ComponentStory<typeof Button> = args => <Button {...args} />

export const Default = Template.bind({});
Default.args = {
  colour: "red",
};
```

Then I'm imported the story into a test file:

```typescript
// Button.test.jsx

import { Default } from "./Button.stories";
import userEvent from "@testing-library/user-event";

it("click the button", async () => {
  render(<Default {...Default.args} />);

  expect(screen.getByRole("button", { name: /a button/i })).toHaveAccessibleName();
  await userEvent.click(screen.getByRole("button"));
});
```

I get the common `could not find react-redux context value; please ensure the component is wrapped in a <Provider>`. 

However, if I set the story up like this: 

```typescript
// Button.stories.tsx

export default {
  title: "Button",
  component: Button,
} as ComponentMeta<typeof Button>;

const Template: ComponentStory<typeof Button> = args => {
        <Provider store={getStore(getInitialState() as State)}>
           <Button {...args} />
       </Provider>
}

export const Default = Template.bind({});
Default.args = {
  colour: "red",
};
```

It works? Why?

This component is a contrived example but we have more complex use cases where keeping our stories dry and decorators in one place makes a lot of sense.

**Comment by Marklb** [comment]
It looks like you are using a story, but not composing it with the `Meta`, so it only knows what is on the imported object. The docs explain it here: https://storybook.js.org/docs/writing-tests/stories-in-unit-tests#write-a-test-with-testing-library

I think you want the following:

```tsx
import Meta, { Default as DefaultStory } from "./Button.stories";
import userEvent from "@testing-library/user-event";
import { composeStory } from '@storybook/react';

const Default = composeStory(DefaultStory, Meta);

it("click the button", async () => {
  render(<Default />);

  expect(screen.getByRole("button", { name: /a button/i })).toHaveAccessibleName();
  await userEvent.click(screen.getByRole("button"));
});
```

**Reply by daniel-norris** [reply]
Hey, thanks for the reply. I should have mentioned this is on a legacy codebase and we're on SB6.4 at the moment. Can't upgrade for various reasons immediately. 

Is there a way to do this on 6? 

https://storybook.js.org/docs/6.4/writing-tests/importing-stories-in-tests

**Reply by Marklb** [reply, marked as answer]
I am not sure why the docs doesn't mention the compose methods, but they were in a separate project in v6. The way they are used should be similar, so I think you would just need to replace `@storybook/react` with `@storybook/testing-react`, for the test functions. https://storybook.js.org/addons/@storybook/testing-react

I mainly work with Angular, so I don't have a SB6.4 React project ready to confirm at the moment. I wrote `@storybook/testing-angular`, based off of the React implementation, and it has SB6.3.8 as dependencies, so the compose methods should be there for React in 6.4.

**Reply by daniel-norris** [reply]
Amazing, this is what I was looking for! thanks.
